Cython/Compiler/Nodes.py

```diff
@@ -3018,21 +3041,49 @@ def analyse_signature(self, env):
             sig.is_staticmethod = True
             sig.has_generic_args = True
 
-        if ((self.is_classmethod or self.is_staticmethod) and
-                self.has_fused_arguments and env.is_c_class_scope):
-            del self.decorator_indirection.stats[:]
-
         for i in range(min(nfixed, len(self.args))):
             arg = self.args[i]
             arg.is_generic = 0
-            if sig.is_self_arg(i) and not self.is_staticmethod:
+            if (sig.is_self_arg(i) and not self.is_staticmethod):
+                self_type_overridden = arg.type_set_manually
                 if self.is_classmethod:
                     arg.is_type_arg = 1
-                    arg.hdr_type = arg.type = Builtin.type_type
+                    usual_type = Builtin.type_type
+                    if not arg.type_set_manually:
+                        arg.hdr_type = arg.type = usual_type
                 else:
                     arg.is_self_arg = 1
-                    arg.hdr_type = arg.type = env.parent_type
+                    if self.is_in_arbitrary_decorator:
+                        usual_type = PyrexTypes.FusedType([env.parent_type, PyrexTypes.py_object_type],
+                                                name="fused self or object")
+                        warn_once(arg.pos, "Type of argument '%s' cannot be assumed to be %s because "
+                                    "it has an unknown decorator. "
+                                    "Consider setting the type explicitly." % (
+                                        arg.name, env.parent_type),
+                                    1, warn_once_key="unknown decorator type")
```

Comment:
Should this really be "warn once"? Shouldn't this be reported at least for each each extension type to which it applies? What would be wrong with always reporting it when it occurs?